### ISSUE TITLE
`limited_fifo_queue_t` uses a `adjustable_semaphore_t`, violating a NO_CORO_WAITING

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#3766]
On `next`, ran a JS test using:

```
r.expr(1).run(conn, callback);
r.expr(1).run(conn, callback);
r.expr(1).run(conn, callback);
r.expr(1).run(conn, callback);
r.expr(1).run(conn, callback);
...
```
The idea being to cause a lot of concurrent queries and weed out any race conditions.

Got the following crash:
```
Version: rethinkdb 1.16.0-1-95-g6f9f6d (debug) (GCC 4.6.3)
error: Error in ../src/arch/runtime/coroutines.cc at line 289:
error: Assertion failed: [TLS_get_cglobals()->assert_no_coro_waiting_counter == 0] This code path is not supposed to use notify_now_deprecated() or spawn_now_dangerously().
error: Backtrace:
error: Wed Feb 11 15:35:21 2015

       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:203
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: coro_t::notify_now_deprecated() at coroutines.cc:288
       7: one_waiter_cond_t::pulse() at cond_var.cc:26
       8: ../build/debug/rethinkdb() [0x167cd6e] at 0x167cd6e ()
       9: adjustable_semaphore_t::lock_request_t::on_available() at semaphore.hpp:92
       10: adjustable_semaphore_t::pump() at semaphore.cc:185
       11: adjustable_semaphore_t::unlock(long) at semaphore.cc:145
       12: limited_fifo_queue_t<ql::protob_t<Query>, std::list<ql::protob_t<Query>, std::allocator<ql::protob_t<Query> > > >::produce_next_value() at limited_fifo.hpp:53
       13: passive_producer_t<ql::protob_t<Query> >::pop() at passive_producer.hpp:83
       14: coro_pool_t<ql::protob_t<Query> >::worker_run(ql::protob_t<Query>, auto_drainer_t::lock_t) at coro_pool.hpp:70
       15: std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)>::operator()(coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t) const at functional:551
       16: void std::_Bind<std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)> (coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2>(std::tuple<>&&, std::_Index_tuple<0, 1, 2>) at functional:1146
       17: void std::_Bind<std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)> (coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1206
       18: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)> (coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       19: callable_action_wrapper_t::run() at runtime_utils.cc:43
       20: coro_t::run() at coroutines.cc:207
       21: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)> (coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (coro_pool_t<ql::protob_t<Query> >::*)(ql::protob_t<Query>, auto_drainer_t::lock_t)> (coro_pool_t<ql::protob_t<Query> >*, ql::protob_t<Query>, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
       22: coro_pool_t<ql::protob_t<Query> >::on_source_availability_changed() at coro_pool.hpp:86
       23: availability_control_t::set_available(bool) at passive_producer.hpp:66
       24: limited_fifo_queue_t<ql::protob_t<Query>, std::list<ql::protob_t<Query>, std::allocator<ql::protob_t<Query> > > >::push(ql::protob_t<Query> const&) at limited_fifo.hpp:35
       25: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, signal_t*, client_context_t*) at protob.cc:390
       26: query_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at protob.cc:311
```

It appears that `limited_fifo_queue_t` inherits from `passive_producer_t`.  `passive_producer_t::pop()` has an `ASSERT_NO_CORO_WAITING`.  This is a problem, because `limited_fifo_queue_t` uses `adjustable_semaphore_t`, which uses a `one_waiter_cond_t`, which uses `coro_t::notify_now_deprecated()`.

It seems it would be easy to change `one_waiter_cond_t` to use `notify_sometime`.  However, I'm hesitant to change any of the underlying synchronization primitives because they're used in so many places, some of which may depend on the current behavior.  `limited_fifo_queue_t` is only used in the `connection_loop` function for ReQL TCP connections, and maybe can be fixed without disrupting anything else.  Ideally, `limited_fifo_queue_t` would use `new_semaphore_t` instead of `adjustable_semaphore_t`, although that raises some RAII problems.

**Comment by Tryneus** [comment]
I tried changing `one_waiter_cond_t` to use `notify_sometime` instead of `notify_now_deprecated` - it was only being used in `static_semaphore_t::co_lock()` and `adjustable_semaphore_t::co_lock()`, both of which would perform a `coro_t::yield_ordered()` right after the `one_waiter_cond_t::wait()`.  I believe this means that users of the `semaphore_t`s already could not count on running immediately, and it is safe to use `notify_sometime` instead and drop the `coro_t::yield_ordered()`.

I am not positive about this reasoning, but with this change, all the unittests and reql tests pass.  I am also running a few valgrind tests to see if anything suspicious shows up.

**Comment by Tryneus** [comment]
The fix for this has been approved and merged to `next` in commit e0ee597.  Review was 2611.  Will be in release 2.0.
